uSSL:  Add compile option for a finaliser

MrSurly opened the pull request:
Per https://github.com/micropython/micropython-esp32/issues/209#issuecomment-339855157

MrSurly committed:
extmod/modussl_mbedtls.c: Added finaliser
Per the comment found here
https://github.com/micropython/micropython-esp32/issues/209#issuecomment-339855157,
adding finaliser code to prevent memory leaks from USSL objects.  This
is in-line with the finaliser code found in many modsocket
implementations for various ports.

MrSurly commented:
If accepted, please propagate to ESP32.

dpgeorge commented:
A problem that this brings up is that the GC is not reentrant.  That means a finaliser cannot explicitly free memory on the uPy heap.  In the best case a call to gc_free() in the finaliser does nothing.  In the worst case (eg with threading enabled) it leads to deadlocks.  That hasn't been an issue in the past because all uses of finalisers are to close files/sockets, not to free memory on the uPy heap (which is anyway done automatically).

The concern then is that for ports that use the uPy heap for allocating mbedtls objects (which is not the case for esp32 which has a separate heap), this patch will lead to bugs.

pfalcon reviewed (changes requested):
Please add to the commit message a detailed description why the change is made and what it brings. (This generally applies to all commits.)

MrSurly commented:
@pfalcon Amended commit message

MrSurly commented:
@dpgeorge Added a guard `#ifdef` around finaliser.

pfalcon reviewed (approved):
Thanks.

pfalcon commented:
Two commits can be squashed into one btw.

pfalcon reviewed (changes requested):
Oh, btw, all options added should be listed in py/mpconfig.h, or we would have lost track of them long ago.

MrSurly commented:
@pfalcon 
* Updated py/mpconfig.h
* Generalized the define to `MICROPY_PY_USSL_FINALISER`
* Added this define, with a comment to `py/mpconfig.h`
* Squashed into one commit

pfalcon commented on py/mpconfig.h:
```diff
@@ -1109,6 +1109,8 @@ typedef double mp_float_t;
 
 #ifndef MICROPY_PY_USSL
 #define MICROPY_PY_USSL (0)
+// Whether to add finaliser code to uSSL
```
Better to say "Whether to add finaliser to ussl objects".

pfalcon reviewed (approved):
Looks good, thanks. Minor documenting comment suggestion.

MrSurly committed:
py/mpconfig.h: Minor comment change

dpgeorge commented:
Merged in 74ec52d85758ad9da7a3abb24257511b22d74964
